frontend/catalyst/pennylane_extensions.py

```diff
@@ -162,54 +181,108 @@ class QJITDevice(qml.QubitDevice):
-        "QubitUnitary",
-        "Adjoint",
-    ]
-    observables = [
-        "Identity",
-        "PauliX",
-        "PauliY",
-        "PauliZ",
-        "Hadamard",
-        "Hermitian",
-        "Hamiltonian",
-    ]
+
+    # These must be present even if empty.
+    operations = []
+    observables = []
+
+    @staticmethod
+    def _get_operations_available_in_QIR_abstract_machine():
+        # These are the methods implemented in the runtime.
+        # A device might have gates which are native to the device
+        # but are not exposed. This is a limitation of the current runtime
+        # because it does not contain implementation of gates for every possible
+        # gate from the infinite set.
+        return {
+            "Identity",
+            "PauliX",
+            "PauliY",
+            "PauliZ",
+            "Hadamard",
+            "S",
+            "T",
+            "PhaseShift",
+            "RX",
+            "RY",
+            "RZ",
+            "Rot",
+            "CNOT",
+            "CY",
+            "CZ",
+            "SWAP",
+            "IsingXX",
+            "IsingYY",
+            "IsingXY",
+            "ControlledPhaseShift",
+            "CRX",
+            "CRY",
+            "CRZ",
+            "CRot",
+            "CSWAP",
+            "Toffoli",
+            "MultiRZ",
+            "QubitUnitary",
+        }
+
+    @staticmethod
+    def _get_operations_to_convert_to_matrix(_config):
+        # We currently override and only set MultiControlledX to preserve current behaviour.
+        # We could choose to read from config and use the "matrix" gates.
+        # However, that affects differentiability.
+        # None of the "matrix" gates with more than 2 qubits parameters are differentiable.
+        # TODO: https://github.com/PennyLaneAI/catalyst/issues/398
+        return {"MultiControlledX"}
+
+    @staticmethod
+    def _check_mid_circuit_measurement(config):
+        return config["compilation"]["mid_circuit_measurement"]
+
+    @staticmethod
+    def _check_adjoint(config):
+        return config["compilation"]["adjoint"]
+
+    @staticmethod
+    def _check_quantum_control(config):
+        return config["compilation"]["quantum_control"]
+
+    @staticmethod
+    def _set_supported_operations(config):
+        """Override the set of supported operations."""
+        native_gates = set(config["operations"]["gates"][0]["native"])
+        qir_gates = QJITDevice._get_operations_available_in_QIR_abstract_machine()
+        QJITDevice.operations = list(native_gates.intersection(qir_gates))
+
+        # These are added unconditionally.
+        QJITDevice.operations += ["Cond", "WhileLoop", "ForLoop"]
+
+        if QJITDevice._check_mid_circuit_measurement(config):
+            QJITDevice.operations += ["MidCircuitMeasure"]
+
+        if QJITDevice._check_adjoint(config):
+            QJITDevice.operations += ["Adjoint"]
+
+        if QJITDevice._check_quantum_control(config):  # pragma: nocover
```

Comment:
How come this is nocover? Surely the if statement gets executed? If you are looking to only disable coverage for one of the branches, there is also `# pragma: nobranch`

Reply:
Ah thanks. I didn't know about `nobranch`. We only have partial coverage of this branch. Thanks!

Reply:
https://github.com/PennyLaneAI/catalyst/pull/369/commits/ab9e86de3498c944ddfc0e8eed8a925678e6a4f5

Reply:
Ah I just realized, if it's the "then" branch that is not covered (rather than the "else" branch as is often the case), you might still need to mark the body as nocover. This instance might not be the best use of `nobranch` then, sorry  about that. (still I think useful to know about)

Reply:
Revert: https://github.com/PennyLaneAI/catalyst/pull/369/commits/1012d94848058e2c3db82ebc42a1e17bef441107